spec/std/http/server/handlers/static_file_handler_spec.cr

```diff
@@ -21,29 +21,94 @@ describe HTTP::StaticFileHandler do
+      headers["If-Modified-Since"] = initial_response.headers["Last-Modified"]
+
       response = handle HTTP::Request.new("GET", "/test.txt", headers), ignore_body: true
       response.status_code.should eq(304)
-      response.headers["Last-Modified"].should eq(HTTP.rfc1123_date(File.info("#{__DIR__}/static/test.txt").modification_time))
+    end
+
+    it "should not return Content-Type if file mtime is equal" do
+      initial_response = handle HTTP::Request.new("GET", "/test.txt")
+
+      headers = HTTP::Headers.new
+      headers["If-Modified-Since"] = initial_response.headers["Last-Modified"]
+
+      response = handle HTTP::Request.new("GET", "/test.txt", headers), ignore_body: true
+      response.headers["Content-Type"]?.should eq(nil)
     end
 
     it "should return 304 Not Modified if file mtime is older" do
       headers = HTTP::Headers.new
       headers["If-Modified-Since"] = HTTP.rfc1123_date(File.info("#{__DIR__}/static/test.txt").modification_time + 1.hour)
       response = handle HTTP::Request.new("GET", "/test.txt", headers), ignore_body: true
+
       response.status_code.should eq(304)
-      response.headers["Last-Modified"].should eq(HTTP.rfc1123_date(File.info("#{__DIR__}/static/test.txt").modification_time))
     end
 
     it "should serve file if file mtime is younger" do
       headers = HTTP::Headers.new
       headers["If-Modified-Since"] = HTTP.rfc1123_date(File.info("#{__DIR__}/static/test.txt").modification_time - 1.hour)
-      response = handle HTTP::Request.new("GET", "/test.txt")
+      response = handle HTTP::Request.new("GET", "/test.txt", headers)
+
+      response.status_code.should eq(200)
+      response.body.should eq(File.read("#{__DIR__}/static/test.txt"))
+    end
+  end
+
+  context "with If-None-Match header" do
+    it "should return 304 Not Modified if header matches etag" do
+      initial_response = handle HTTP::Request.new("GET", "/test.txt")
+
+      headers = HTTP::Headers.new
+      headers["If-None-Match"] = initial_response.headers["Etag"]
+      response = handle HTTP::Request.new("GET", "/test.txt", headers), ignore_body: true
+      response.status_code.should eq(304)
+    end
+
+    it "should serve file if header does not match etag" do
+      headers = HTTP::Headers.new
+      headers["If-None-Match"] = "some random etag"
+
+      response = handle HTTP::Request.new("GET", "/test.txt", headers)
+      response.status_code.should eq(200)
+      response.body.should eq(File.read("#{__DIR__}/static/test.txt"))
+    end
+  end
+
+  context "with both If-None-Match and If-Modified-Since headers" do
+    it "ignores If-Modified-Since as specified in RFC 7232" do
+      initial_response = handle HTTP::Request.new("GET", "/test.txt")
+
+      headers = HTTP::Headers.new
+      headers["If-Modified-Since"] = HTTP.rfc1123_date(File.info("#{__DIR__}/static/test.txt").modification_time - 1.hour)
```

Comment:
This would also benefit from using `Last-Modified`: 

```cr
HTTP.parse_time(initial_response.headers["Last-Modified"]) - 1.hour
```

Reply:
but you would have to parse it back with `rfc1123_date` (not mentioning that `parse_time` returns `Time?`, so `not_nil!` would have to be added as well) - is it really worth it? 🤔